onnxruntime/contrib_ops/cpu/bert/multihead_attention.cc

```diff
@@ -0,0 +1,464 @@
+      [](BroadcastHelper& per_iter_bh) {
+        per_iter_bh.OutputEigen<T>() = per_iter_bh.EigenInput0<T>() + per_iter_bh.EigenInput1<T>();
+      }}; // For element-wise add
+
+  // Get Q's bias from combined bias or QKV's bias from combined bias
+  AllocatorPtr allocator;
+  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&allocator));
+  std::vector<int64_t> bias_dims({num_heads * head_size});
+  if (packed_qkv) {
+    bias_dims = {3 * num_heads * head_size};
+  }
+  gsl::span<const int64_t> bias_dims_span{bias_dims};
+  TensorShape bias_shape(bias_dims_span);
+  OrtValue bias;
+  Tensor::InitOrtValue(element_type, bias_shape, allocator, bias);
+  auto num_bias_elements = packed_qkv ? 3 * hidden_size * element_size : hidden_size * element_size;
+  memcpy(bias.GetMutable<Tensor>()->MutableData<T>(), qkv_bias + bias_offset, num_bias_elements);
+  
+  // Compute Q(BS, D) + bias(D) or QKV(BS, 3*D) + bias(3*D) as broadcasted element-wise add
+  {
+    InputBroadcaster input_broadcaster(*bias.GetMutable<Tensor>(), *qkv);
+    const InputBroadcaster& const_input_broadcaster = input_broadcaster;
+    Tensor& output_tensor = *qkv_with_bias.GetMutable<Tensor>();
+
+    size_t span_size = input_broadcaster.GetSpanSize();
+    size_t output_size = static_cast<ptrdiff_t>(output_tensor.Shape().Size());
+    void* user_data = nullptr;
+
+    const int loop_len = static_cast<int>(output_size / span_size);
+    double unit_cost = 1.0f;
+    const auto cost = TensorOpCost{static_cast<double>(input_broadcaster.Input0ElementSize()) * span_size,
+                                   static_cast<double>(output_tensor.DataType()->Size()) * span_size,
+                                   unit_cost * span_size};
+    auto tp = context->GetOperatorThreadPool();
+    ThreadPool::TryParallelFor(tp, loop_len, cost, \
+      [span_size, &const_input_broadcaster, &output_tensor, &add_funcs, user_data](std::ptrdiff_t first_span,
+                                                                                   std::ptrdiff_t last_span) {
+        InputBroadcaster segment_input_broadcaster(const_input_broadcaster);
+        segment_input_broadcaster.AdvanceBy(first_span * span_size);
+
+        OutputBroadcaster segment_output_broadcaster(span_size, output_tensor,
+                                                     first_span * span_size, last_span * span_size);
+
+        BroadcastHelper segment_helper(segment_input_broadcaster, segment_output_broadcaster, user_data);
+        BroadcastLooper(segment_helper, add_funcs);
+    });
+  }
+
+  // Reshape Q from BxSxD to BxNxSxH or QKV from BxSx(3xD) to 3xBxNxSxH
+  std::vector<int64_t> reshape_dims({batch_size, num_heads, sequence_length, head_size});
+  if (packed_qkv) {
+    reshape_dims = {3, batch_size, num_heads, sequence_length, head_size};
+  }
+  gsl::span<const int64_t> reshape_dims_span{reshape_dims};
+  TensorShape qkv_final_dims(reshape_dims_span);
+  qkv_with_bias.GetMutable<Tensor>()->Reshape(qkv_final_dims);
```

Comment:
Does it need to reshape here since memory layout is not changed?.

Reply:
Reshape is called to update the tensor's shape property. According to the [comments](https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/framework/tensor.h#L244-L254), no memory layout is changed when this is called.